### DISCUSSION TITLE
Data received from serial is cut off

### DISCUSSION THREAD

**Discussion by Mrcubix** [hathach/tinyusb#1184] [Q&A] [answered]
Device: rpi Pico (rp2040)

I'm trying to send data over USB serial using tinyUSB to debug some of my problems.
The issue is that the output string is cut off, it should appear as follow:

```txt
Connected

Test 1:
1621061-GPIO28-0
1621245-GPIO22-0

Test 2:
1621341-GPIO28-1
1621372-GPIO22-0

Test 3:
1621440-GPIO28-1
1621471-GPIO22-1
```
but what i receive is:
```txt
Connected

Test 1:
1621061-GPIO28-0
1621245-GPIO22-0

Test 2:
1621341-GPIO28-1
1621372-GPIO22-0

Test 3:
1621440-
```
An here the part of my code in main responsible for the previous output:
<details>
    <summary>Code</summary>
    
```c
// Invoked when cdc when line state changed e.g connected/disconnected
void tud_cdc_line_state_cb(uint8_t itf, bool dtr, bool rts)
{
    (void) itf;
    (void) rts;

    if ( dtr )
    {
        tud_cdc_write_str("Connected\r\n");

        Test1();

        tud_cdc_write_flush();

        Test2();

        tud_cdc_write_flush();

        Test3();

        tud_cdc_write_flush();

    }
}

void CheckPins()
{
    // Send the GPIO n°key0SensePin's state
    PrepareString(time_us_64(), key0SensePin, GetPin(key0SensePin));
    serial_send();
    // Send the GPIO n°key0SendPin's state
    PrepareString(time_us_64(), key0SendPin, GetPin(key0SendPin));
    serial_send();
}

void Test1()
{
    serial_print_separator();
    serial_print("Test 1:\n\r");

    CheckPins();

    serial_print_separator();
}

void Test2()
{
    serial_print("Test 2:\n\r");

    // Drive the GPIO n°key0SensePin's HIGH
    gpio_put(key0SensePin, 1);
    // Drive the GPIO n°key0SendPin's HIGH
    gpio_put(key0SendPin, 1);
    // Send the GPIO n°key0SensePin's state
    CheckPins();

    serial_print_separator();
}

void Test3()
{
    serial_print("Test 3:\n\r");

    InverseDir();
    CheckPins();

    serial_print_separator();
}
``` 
<sub>Here, basing myself on some things i read around online (which is probably wrong) `tud_cdc_write_flush()` is normally used to flush data off tinyusb?</sub>
</details>

It seem to be related to tinyusb's buffer as commenting one of my test in my code involving sending data will make the remaining data appear in the terminal (PuTTY).

My first guess would be that tinyusb's buffer is full, and that it isn't being cleared, which would mean that i'm unable to send more.

Unfortunately, i cannot take appropriate steps to diagnose this issue as i lack experience with how tinyusb work
(tinyusb's lack of documentation doesn't help either).

I'm pretty sure i can fix this on my own provided someone guided me in the troubleshooting process.

**Comment by hathach** [comment, marked as answer]
The buffering scheme for cdc may sound complicated but is very simply. 
- cdc write() will put application contents into an cdc's fifo
- tinyusb will send data to usb bus (flush) when enough data is buffered, mostly 64 bytes for fullspeed and 512 bytes for highspeed.
- flush() will manually force data (even less than 64 bytes) to be sent to bus immediately.
- Each bus transfer typically take 1 ms for FS, 125 us for HS.

back to your code, 1st flush() will sent data to bus, 2nd and 3rd flush() will return 0 since the usb bus is busy already. Data remains in the fifo, keeping bumping data will overflow the fifo. You should either add check to see if cdc is ready or write_available(), or increase fifo.

**Reply by Mrcubix** [reply]
Thanks for the answer, it seem like i'll have to somehow wait between tests to avoid this issue

EDIT:

i guess something like that will do for my use case, increasing the buffer size in `tusb_config.h` doesn't seem to affect the fifo size:

<details>
    <summary>Code</summary>

```c
static int testIndex = 1;
void cdc_task();

int main(void)
{
    board_init(); // Initialize stuff such as led?
    tusb_init(); // initialize the device
    
    board_led_write(true);

    Setup();

    while (1)
    {
        tud_task(); // tinyusb device task, handle event such as mount and unmout, used to provide the descriptor and configs
        cdc_task();
    }

    return 0;
}

// Invoked when cdc when line state changed e.g connected/disconnected
void tud_cdc_line_state_cb(uint8_t itf, bool dtr, bool rts)
{
    (void) itf;
    (void) rts;

    if ( dtr )
    {
        tud_cdc_write_str("Connected\r\n");
        tud_cdc_write_flush();
    }
    else
    {
        testIndex = 1;
    }
}

void cdc_task()
{
    if (tud_cdc_connected() && tud_cdc_write_available() == 64)
    {
        switch(testIndex)
        {
            case 1:
                Test1();
                break;
            case 2:
                Test2();
                break;
            case 3:
                Test3();
                break;
        }

        tud_cdc_write_flush();

        if (testIndex != 4)
            testIndex++;
    }
}
```
</details>

output:

```txt
Connected

Test 1:
3024588-GPIO28-0
3024771-GPIO22-0

Test 2:
3024860-GPIO28-1
3024900-GPIO22-0

Test 3:
3024985-GPIO28-1
3025029-GPIO22-1
```

**Reply by hathach** [reply]
Which macro you increase?
